src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+            if (r.error != error_code::none) {
+                errored_partitions.push_back(i);
+            }
+        }
+
+        if (
+          total_size > ctx.min_bytes || !errored_partitions.empty()
+          || ctx.as.abort_requested()) {
+            co_return worker_result{
+              .read_results = std::move(results),
+              .errored_partitions = std::move(errored_partitions),
+              .total_size = total_size,
+            };
+        }
+
+        // If the fast-path doesn't return there isn't enough data so
+        // we'll have to wait for the last_visible_index of partitions to
+        // change.
+        ss::gate waiter_gate;
+        ss::condition_variable has_requests;
+        std::vector<size_t> request_indices;
+
+        // An abort source is needed to clear any waiters from the
+        // `raft::offset_monitor` after the worker returns.
+        ss::abort_source waiter_as;
+        auto hwm_as_sub = _as.subscribe([&waiter_as, &has_requests]() noexcept {
+            waiter_as.request_abort();
+            // ss::condition_variable doesn't take an abort source when waiting,
+            // so just signal it to wake up on aborts.
+            has_requests.signal();
+        });
+
+        for (size_t i = 0; i < ctx.requests.size(); i++) {
+            auto wait_opt = register_hwm_wait(
+              ctx,
+              ctx.requests[i].ktp(),
+              model::next_offset(last_visible_indexes[i]),
+              waiter_as);
+
+            if (!wait_opt) {
+                results[i] = read_result(error_code::not_leader_for_partition);
+                results[i].partition = ctx.requests[i].ktp().get_partition();
+                errored_partitions.push_back(i);
+                break;
+            }
+
+            ssx::spawn_with_gate(
+              waiter_gate,
+              [w = std::move(wait_opt.value()),
+               &request_indices,
+               &has_requests,
+               i]() mutable {
+                  return w.finally([&request_indices, &has_requests, i] {
+                      request_indices.push_back(i);
+                      has_requests.signal();
+                  });
+              });
+        }
+
+        if (!errored_partitions.empty()) {
+            waiter_as.request_abort();
+            co_await waiter_gate.close();
+
+            co_return worker_result{
+              .read_results = std::move(results),
+              .errored_partitions = std::move(errored_partitions),
+              .total_size = total_size,
+            };
+        }
+
+        std::vector<size_t> results_map;
+
+        for (;;) {
+            co_await has_requests.wait();
+
+            if (_as.local().abort_requested()) {
+                break;
+            }
+
+            results_map.clear();
+            requests.clear();
+
+            // Only query the partitions that had a HWM increase.
+            // Its important not to introduce suspension points in the loop
```

Comment:
Rather than writing it is as a comment you can move the part that shouldn't be suspended into a non-async function which will enforce this.

Reply:
Good point, wonder if we should codify this a bit more and have a library function(`utils::atomic_lambda`, `utils::no_async`...) that takes in a non-async closure.